envoy/router/router.h

```diff
@@ -1075,6 +1075,15 @@ class Route {
   template <class Derived> const Derived* perFilterConfigTyped(const std::string& name) const {
     return dynamic_cast<const Derived*>(perFilterConfig(name));
   }
+
+  /**
+   * @return const envoy::config::core::v3::Metadata& return the metadata provided in the config for
+   * this route.
+   */
+  virtual const envoy::config::core::v3::Metadata& metadata() const { return empty_metadata_; }
```

Comment:
Do you know why we don't just return a RouteEntry for direct response entries? I'm a bit concerned that exposing metadata on both Route and RouteEntry leaves the API in an odd state where it's not clear why one would use on over the other. Thoughts?

Reply:
yeah， it will be odd state. but how can it get the metadata int http filter for direct response entries?

Reply:
I think it probably would be populated on the `RouteEntryImplBase` already, we just somehow refuse to give the caller a RouteEntry when it's a direct response. @mattklein123 @alyssawilk any idea why we special case direct response in this way? 

The current proposed implementation seems incomplete as well, seems like we'd always return empty metadata in this case?

Reply:
yeah seems weird to me, I'd rather (runtime guard) expose the route entry on direct response but maybe git blame who special cased it and ask them?

Reply:
From looking at the code, I suspect the original intention is that a direction response != an upstream route entry, and that is why the interface/code is broken down that way. I have 2 comments:
1) Personally, I would remove metadata() from RouteEntry and add it to Route as this PR does. Then it can be grabbed in all cases.
2) I don't understand how always returning empty metadata here makes sense. Don't you want to return the configured route metadata?

Reply:
yeah, if metadata() interface in RouteEntry, it will make many weird problems for direct response entries. ok, i will remove metadata() from RouteEntry.

Reply:
i try to remove metadata() from Router::RouteEntry, so many files will be change. i think it will take a another problem.

Reply:
Yes, it will be a larger change, but that is the change we need to do.

Reply:
ok， i will do it

Reply:
I'm still not sure why this is returning empty metadata. I would keep this as a PURE virtual function and let the implementation return the right thing.

/wait